libs/Microsoft.MixedReality.WebRTC.Unity/Assets/Microsoft.MixedReality.WebRTC.Unity/Scripts/PeerConnection.cs

```diff
@@ -1,608 +1,1208 @@
+        /// renegotiating some new or removed tracks.
+        /// </summary>
+        public bool CreateOffer()
+        {
+            if (Peer == null)
+            {
+                throw new InvalidOperationException("Cannot create an offer with an uninitialized peer.");
+            }
+
+            // Batch all changes into a single offer
+            AutoCreateOfferOnRenegotiationNeeded = false;
+
+            // Add all new transceivers for local tracks. Since transceivers are only paired by negotiated mid,
+            // we need to know which peer sends the offer before adding the transceivers on the offering side only,
+            // and then pair them on the receiving side. Otherwise they are duplicated, as the transceiver mid from
+            // locally-created transceivers is not negotiated yet, so ApplyRemoteDescriptionAsync() won't be able
+            // to find them and will re-create a new set of transceivers, leading to duplicates.
+            // So we wait until we know this peer is the offering side, and add transceivers to it right before
+            // creating an offer. The remote peer will then match the transceivers by mlineIndex, then add any missing
+            // ones, after it applied the offer.
+            {
+                // In case this is not the first offer, some transceivers might already exist
+                var transceivers = _nativePeer.Transceivers;
+                int numTransceivers = transceivers.Count;
+
+                for (int mlineIndex = 0; mlineIndex < _mediaLines.Count; ++mlineIndex)
+                {
+                    var mediaLine = _mediaLines[mlineIndex];
+
+                    // Compute the transceiver desired direction based on what the local peer expects, both in terms
+                    // of sending and in terms of receiving. Note that this means the remote peer will not be able to
+                    // send any data if the local peer did not add a remote source first.
+                    // Tracks are not tested explicitly since the local track can be swapped on-the-fly without renegotiation,
+                    // and the remote track is generally not added yet at the beginning of the negotiation, but only when
+                    // the remote description is applied (so for the offering side, at the end of the exchange when the
+                    // answer is received).
+                    bool wantsSend = (mediaLine.Sender != null);
+                    bool wantsRecv = (mediaLine.Receiver != null);
+                    var wantsDir = (wantsSend ? (wantsRecv ? Transceiver.Direction.SendReceive : Transceiver.Direction.SendOnly)
+                        : (wantsRecv ? Transceiver.Direction.ReceiveOnly : Transceiver.Direction.Inactive));
+
+                    // Ensure the media line has a transceiver, and update its desired direction
+                    Transceiver tr = null;
+                    if (mlineIndex < transceivers.Count)
```

Comment:
nit: IMHO it would be nicer to split into two loops:
```
for (int mlineIndex = 0; mlineIndex < transceivers.Count; ++mlineIndex) { ... }
for (int mlineIndex = transceivers.Count; mlineIndex < _mediaLines.Count; ++mlineIndex) { ... }
```
(transceivers are always <= media lines, right?) 
This way the part that goes
```
tr = _nativePeer.AddTransceiver(mediaLine.Kind, settings);
...
Debug.Assert(tr.MlineIndex == mlineIndex);
```
becomes a bit clearer because it's evident that the new transceivers are all sequential.

Reply:
Fixed.